site/zenodo_rdm/exporter/tasks.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2025 CERN.
+#
+# ZenodoRDM is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""ZenodoRDM exporter tasks."""
+
+import gzip
+import io
+import json
+from datetime import datetime
+
+from celery import shared_task
+from flask import current_app
+from invenio_app.factory import create_api
+from invenio_communities.communities.records.models import CommunityMetadata
+from invenio_db import db
+from invenio_files_rest.models import Bucket, Location, ObjectVersion, as_bucket
+from invenio_search.proxies import current_search_client as client
+from opensearch_dsl import Search
+
+from zenodo_rdm.exporter.config import (
+    EXPORTER_BUCKET_UUID,
+    EXPORTER_NUMBER_VERSIONS_TO_KEEP,
+)
+
+
+def _dump_records(community_slug):
+
+    community_uuid = (
+        db.session.query(CommunityMetadata.id)
+        .filter(CommunityMetadata.slug == community_slug)
+        .scalar()
+    )
+
+    app = create_api()
+
+    with app.app_context():
+        idxprefix = current_app.config["SEARCH_INDEX_PREFIX"]
+
+        # Dump the records
+        t1 = datetime.now()
+        print("Writing records ...")
+        query = f"is_published:true AND is_deleted:false AND parent.communities.ids:{community_uuid}"
+        s = (
+            Search(using=client, index=f"{idxprefix}rdmrecords-records")
+            .extra(
+                track_total_hits=True,
+            )
+            .query(
+                "query_string",
+                query=query,
+            )
+        )
+
+        stream = io.StringIO()
+
+        i = 0
+        for r in s.scan():
+            stream.write(json.dumps(r._d_))
+            i += 1
+            if i % 1000 == 0:
+                print("...", i)
+        print(f"... wrote {i} records")
+
+        t2 = datetime.now()
+        print("Writing records", t2 - t1)
+
+        stream.seek(0)
+
+        return stream
+
+
+def _create_or_get_bucket():
+    bucket = as_bucket(EXPORTER_BUCKET_UUID)
+    if not bucket:
+        print(f"Creating exporter bucket: {EXPORTER_BUCKET_UUID}")
+        with db.session.begin_nested():
+            # TODO: Set `quota_size` and `max_file_size`.
+            bucket = Bucket(
+                id=EXPORTER_BUCKET_UUID,
+                default_location=Location.get_default().id,
+                default_storage_class=current_app.config[
+                    "FILES_REST_DEFAULT_STORAGE_CLASS"
+                ],
+            )
+            db.session.add(bucket)
+            db.session.commit()  # TODO: Commit here OK?
```

Comment:
This should be fine here.

In general with all the objects we create we should make sure that consumers don't see any "transient" state exposed (e.g. an ObjectVersion/FileInstance that was just initialized and is currently being written to). At the same time we should also avoid having a long DB transaction that takes place for the entirety of the exporter process, which might take hours.